### PR TITLE
Implement basic ui for bank interface - Basic ui v2 using rivo tview

### PR DESCRIPTION
**Pull request by koden-km**
Not fully complete, but covers the basics at this stage.
* Customer signup
* Account creation
* Account list
* Deposit
* Withdraw
* Transfer

Refs https://github.com/dogmatiq/example/issues/23

### CODE REVIEW

**Comment by koden-km** [comment]
It really needs some more work to make it better structured, but its functional if you want to try it out in your terminal.

File: cmd/bank/main.go [hunk 1]
```diff
@@ -1,23 +1,121 @@
...
+}
+
 func businessDayFromTime(t time.Time) string {
 	return t.Format(messages.BusinessDateFormat)
 }
 
+func readString(reader *bufio.Reader) string {
+	text, _ := reader.ReadString('\n')
+	return strings.TrimSuffix(text, "\n")
+}
```
Comment (koden-km) [inline cmd/bank/main.go:53]: This function should be removed. Its from the first text UI design and is no longer used.

File: cmd/bank/main.go [hunk 2]
```diff
@@ -1,23 +1,121 @@
...
+	return fmt.Sprintf("%d%04d", rand.Intn(8)+1, rand.Intn(9000))
+}
+
+type appState struct {
+	db     *sql.DB
+	engine *engine.Engine
+	reader *bufio.Reader
+	time   time.Time
+
+	// ui ...
```
Comment (koden-km) [inline cmd/bank/main.go:73]: I doubt that I need to store all these now. Most that are recreated each time could just stay in the scope they are defined in?

File: cmd/bank/main.go [hunk 3]
```diff
@@ -1,23 +1,121 @@
...
+}
+
+func generateAccountNumber() string {
+	return fmt.Sprintf("%d%04d", rand.Intn(8)+1, rand.Intn(9000))
+}
+
+type appState struct {
+	db     *sql.DB
+	engine *engine.Engine
+	reader *bufio.Reader
```
Comment (koden-km) [inline cmd/bank/main.go:70]: This reader should be removed. Its from the first text UI design and is no longer used.

File: cmd/bank/main.go [hunk 4]
```diff
@@ -31,60 +129,619 @@ func main() {
 		panic(err)
 	}
 
-	messages := []dogma.Message{
+	as := appState{
+		db:     db,
+		engine: en,
+		reader: bufio.NewReader(os.Stdin),
```
Comment (koden-km) [inline cmd/bank/main.go:135]: This reader should be removed. Its from the first text UI design and is no longer used.

File: cmd/bank/main.go [hunk 5]
```diff
@@ -31,60 +129,619 @@ func main() {
...
+
+	as.createViews()
+
+	go updateTime(&as)
+
+	if err := as.guiApp.SetRoot(as.layoutView, true).Run(); err != nil {
+		panic(err)
+	}
+
+	fmt.Println(as.infoView.GetText(true))
```
Comment (koden-km) [inline cmd/bank/main.go:147]: I think this was a left over from some debugging and can be removed.

File: cmd/bank/main.go [hunk 6]
```diff
@@ -31,60 +129,619 @@ func main() {
...
-		commands.OpenAccountForNewCustomer{
-			CustomerID:   "cust2",
-			CustomerName: "Bob Jones",
-			AccountID:    "acct2",
-			AccountName:  "Bob Jones",
+		engine.WithCurrentTime(as.time),
+	)
+}
+
+func (as *appState) submitOpenNewAccount(customer customerData, accountName string) {
```
Comment (koden-km) [inline cmd/bank/main.go:586]: This only needs the customer ID, not a full customerData value.

File: cmd/bank/main.go (hunk 2 above)
Comment (koden-km) [reply cmd/bank/main.go:65]: Removed.

**Comment by koden-km** [comment]
What do you do to fix `go.sum` conflicts? There was some `make` command right?

**Comment by jmalloc** [comment]
You can either just resolve the conflict manually -- remember it's not a lock file so you can pretty much just retain both edits usually, or otherwise delete the file and just run `go mod tidy`.

**Comment by jmalloc** [comment]
Closing, we're doing a web UI instead.

**Pull request closed without merging**